lib/printer.js

```diff
@@ -568,9 +568,11 @@ function genericPrintNoParens(path, options, print) {
 
         if (n.argument) {
             var argLines = path.call(print, "argument");
-            if (argLines.length > 1 &&
-                namedTypes.JSXElement &&
-                namedTypes.JSXElement.check(n.argument)) {
+            if (argLines.isFirstLineOnlyComment() ||
```

Comment:
That actually misses some cases. I think you'd need `n.argument.comments`, not `n.comments`, but even then it only gets comments on the top-level node. If, for example, you have a binary operator and a leading comment on the LHS, then the expression would still need to be wrapped in parens. I originally was thinking of doing an AST-based approach like this, but I think it requires repeatedly traversing the leftmost child and checking for any leading comments on that path, which seems non-trivial. Just doing string matching on the source code seems simpler and more robust in a certain sense (since the source code is what the parser is going to see).

I didn't have a test for the LHS binary operator case, but I'll add one.